lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,378 @@
+                                        dtype=np.float32)
+
+        # Attempts to extract realizations. If cube doesn't contain the
+        # dimension then place within list.
+        try:
+            slices_over_realization = temperature_cube.slices_over(
+                "realization")
+        except iris.exceptions.CoordinateNotFoundError:
+            slices_over_realization = [temperature_cube]
+
+        # Creates cube list to hold lapse rate data.
+        lapse_rate_cube_list = iris.cube.CubeList([])
+
+        for temp_slice in slices_over_realization:
+
+            # Create slice to store lapse rate values.
+            lapse_rate_slice = temp_slice
+
+            temperature_data = temp_slice.data
+
+            # Fill sea points with NaN values. Can't use Numpy mask since not
+            # recognised by "generic_filter" function.
+            temperature_data = np.where(land_sea_mask, temperature_data,
+                                        np.nan)
+
+            # Saves all neighbourhoods into "all_temp_subsections".
+            # cval is value given to points outside the array.
+            fnc = SaveNeighbourhood(allbuffers=all_temp_subsections)
+            generic_filter(temperature_data, fnc.filter,
+                           size=self.nbhood_size,
+                           mode='constant', cval=np.nan)
+
+            fnc = SaveNeighbourhood(allbuffers=all_orog_subsections)
+            generic_filter(orography_data, fnc.filter,
+                           size=self.nbhood_size,
+                           mode='constant', cval=np.nan)
+
+            # height_diff_mask is True for points where the height
+            # difference between the central point and its neighbours
+            # is > max_height_diff.
+            height_diff_mask = self._create_heightdiff_mask(
+                all_orog_subsections)
+
+            # Mask points with extreme height differences as NaN.
+            all_orog_subsections = np.where(height_diff_mask, np.nan,
+                                            all_orog_subsections)
+            all_temp_subsections = np.where(height_diff_mask, np.nan,
+                                            all_temp_subsections)
+
+            # Create 1D array for lapse rate and enforce single precision to
+            # speed up calculations.
+            lapse_rate_array = np.zeros(dataarray_size, dtype=np.float32)
+
+            # Loop through both arrays and find gradient of each subsection.
+            # The gradient indicates lapse rate - save into another array.
+            # TODO: This for loop is the bottleneck in the code - 1 array takes
+            # approximately 29 seconds out of a total running time of 33
+            # seconds for UKVX data. This section needs to be parallelised.
+            for i in range(dataarray_size):
+                lapse_rate_array[i] = self._calc_lapse_rate(
+                    all_temp_subsections[i, :], all_orog_subsections[i, :])
```

Comment:
I think you can shave a few seconds off by using list comprehension rather than a loop here.  Eg replace lines 350 - 361 with:
```
lapse_rate_array = [self._calc_lapse_rate(temp, orog)
                                for temp, orog in zip(all_temp_subsections, all_orog_subsections)]
lapse_rate_array = np.array(lapse_rate_array).reshape(dataarray_shape)
```
I've only tested this in a dummy notebook, so you'd want to look at what it does with your UKV data.

Reply:
I've tested this code and your method has a marginal increase in speed (only a second or so) but  I've implemented it.
But there is definitely an argument for changing the _calc_lapse_rate function, we just want to calculate a straight line of fit through 9 data points, so using the least squares regression function is possibly overkill (its designed to fit lines to all types of plots). It might be faster to do the calculation ourselves by applying numpy operations to the whole arrays (which is how the FORTRAN code does it). That way we make use of Numpys optimisation.

Reply:
Good thinking.  Let's discuss on the JIRA ticket :)